sources/packages/backend/libs/services/src/notifications/notification/notification.service.ts

```diff
@@ -70,57 +84,151 @@ export class NotificationService extends RecordDataModelService<Notification> {
   /**
    * Updates the date sent column of the inbox notification record.
    * @param notificationId notification id.
-   * @param entityManager optional repository that can be provided, for instance,
-   * to execute the command as part of an existing transaction. If not provided
-   * the local repository will be used instead.
+   * @param permanentFailureError Error details of a permanent failure if occurred while processing the notification.
    * @returns result of the record updated.
    */
   async updateNotification(
     notificationId: number,
-    entityManager?: EntityManager,
+    permanentFailureError?: PermanentFailureError[],
   ): Promise<UpdateResult> {
-    const notificationRepo =
-      entityManager?.getRepository(Notification) ?? this.repo;
-    return notificationRepo.update(
+    return this.repo.update(
       {
         id: notificationId,
       },
-      { dateSent: new Date() },
+      { dateSent: new Date(), permanentFailureError },
     );
   }
 
   /**
-   * Invokes the sendEmailNotification method of the GC Notification service.
-   * @param notificationId notification id used to retrieve the payload that will be passed
-   * as a parameter to the sendMailNotification method.
-   * @param entityManager optional repository that can be provided, for instance,
-   * to execute the command as part of an existing transaction. If not provided
-   * the local repository will be used instead.
-   * @returns GC Notify API call response.
+   * Process all the unsent notifications with a polling limit.
+   * Processing continues recursively until all the records are processed.
+   * Call GCNotify to send email notification.
+   * @param pollingRecordsLimit Maximum number of notifications to be processed in one chunk of
+   * processing.
+   * @returns processing summary.
    */
-  async sendEmailNotification(
-    notificationId: number,
-    entityManager?: EntityManager,
-  ): Promise<GCNotifyResult> {
-    const notificationRepo =
-      entityManager?.getRepository(Notification) ?? this.repo;
-    const notification = await notificationRepo.findOne({
+  async processUnsentNotifications(
+    pollingRecordsLimit: number,
+  ): Promise<NotificationProcessingSummary> {
+    return await this.processUnsentNotificationsRecursive(pollingRecordsLimit);
+  }
+
+  /**
+   * Call GC Notify to process the given notification.
+   * @param notification notification to be processed.
+   * @returns status of notification processing.
+   */
+  private async sendEmailNotification(
+    notification: Notification,
+  ): Promise<boolean> {
+    // Call GC Notify send email method.
+    try {
+      await this.gcNotifyService.sendEmailNotification(
```

Comment:
Can you please explain, why we do not have a promise.all for this and the updateNotification? I can see if the first call fails, the processNotification is updated with errors in the table. Is it because of that call, its not synced?